src/dataset.js

```diff
@@ -61,11 +62,67 @@ export class Dataset {
      */
     pushData(data) {
         checkParamOrThrow(data, 'data', 'Array | Object');
+        const dispatch = payload => datasets.putItems({ datasetId: this.datasetId, data: payload });
+        const toJSONArray = array => `[${array.join(',')}]`;
+        const checkAndSerialize = (item, index) => {
+            const payload = JSON.stringify(item);
+            const bytes = Buffer.byteLength(payload);
+            if (bytes > MAX_PAYLOAD_SIZE_BYTES) {
+                const s = index ? ` at index ${index} ` : ' ';
+                throw new Error(`Pushed data${s}too large! Actual: ${bytes} bytes. Limit: ${MAX_PAYLOAD_SIZE_BYTES}`);
+            }
+            return { bytes, payload };
+        };
 
-        return datasets.putItems({
-            datasetId: this.datasetId,
-            data,
-        });
+        // handle singular Objects
+        if (!Array.isArray(data)) {
+            try {
+                const { payload } = checkAndSerialize(data);
+                return dispatch(payload);
+            } catch (err) {
+                return Promise.reject(err);
+            }
+        }
+
+        // handle Arrays
+        let items;
+        let totalBytes = 0;
+        try {
+            items = data.map((item, index) => {
+                const piece = checkAndSerialize(item, index);
+                totalBytes += piece.bytes;
+                return piece;
+            });
+        } catch (err) {
+            return Promise.reject(err);
+        }
+
+        // send immediately if not over limit
+        if (totalBytes < MAX_PAYLOAD_SIZE_BYTES) return dispatch(toJSONArray(items.map(i => i.payload)));
+
+        // reducer object
+        const r = { lastChunkBytes: 0, chunks: [] };
```

Comment:
This is very cryptic code, very difficult to read. Can we simplify it a bit? What's the point of having `r` object here?